crates/core/tedge_api/src/workflow/state.rs

```diff
@@ -176,39 +239,86 @@ impl GenericCommandState {
     }
 
     pub fn operation(&self) -> Option<String> {
-        match self.topic.name.split('/').collect::<Vec<&str>>()[..] {
-            [_, _, _, _, _, "cmd", operation, _] => Some(operation.to_string()),
-            _ => None,
-        }
+        extract_command_identifier(&self.topic.name).map(|(operation, _)| operation)
     }
 
     pub fn cmd_id(&self) -> Option<String> {
-        match self.topic.name.split('/').collect::<Vec<&str>>()[..] {
-            [_, _, _, _, _, "cmd", _, cmd_id] => Some(cmd_id.to_string()),
-            _ => None,
-        }
+        extract_command_identifier(&self.topic.name).map(|(_, cmd_id)| cmd_id)
     }
 
-    pub fn is_terminal(&self) -> bool {
-        matches!(self.status.as_str(), "successful" | "failed")
+    pub fn is_successful(&self) -> bool {
+        matches!(self.status.as_str(), SUCCESSFUL)
+    }
+
+    pub fn is_failed(&self) -> bool {
+        matches!(self.status.as_str(), FAILED)
+    }
+}
+
+/// Return the invoking command topic name, if any
+pub fn invoking_command(sub_command: &TopicName) -> Option<TopicName> {
+    match sub_command.split('/').collect::<Vec<&str>>()[..] {
+        [pre, t1, t2, t3, t4, "cmd", _, sub_id] => extract_invoking_command_id(sub_id)
+            .map(|(op, id)| format!("{pre}/{t1}/{t2}/{t3}/{t4}/cmd/{op}/{id}")),
+        _ => None,
+    }
+}
+
+/// Build a sub command identifier from its invoking command identifier
+///
+/// Using such a structure command id for sub commands is key
+/// to retrieve the invoking command of a sub-operation from its state using [extract_invoking_command_id].
+fn sub_command_id(operation: &str, cmd_id: &str) -> String {
+    format!("sub:{operation}:{cmd_id}")
+}
+
+/// Extract the invoking command identifier from a sub command identifier
+///
+/// Return None if the given id is not a sub command identifier
+/// i.e. if not generated with [sub_command_id].
+fn extract_invoking_command_id(sub_cmd_id: &str) -> Option<(&str, &str)> {
+    match sub_cmd_id.split(':').collect::<Vec<&str>>()[..] {
+        ["sub", operation, cmd_id] => Some((operation, cmd_id)),
+        _ => None,
+    }
+}
+
+pub fn extract_command_identifier(topic: &str) -> Option<(String, String)> {
+    match topic.split('/').collect::<Vec<&str>>()[..] {
+        [_, _, _, _, _, "cmd", operation, cmd_id] => {
+            Some((operation.to_string(), cmd_id.to_string()))
+        }
+        _ => None,
```

Comment:
Yeah, this topic root is a true burden ;-)

Reply:
This has been partially fixed by https://github.com/thin-edge/thin-edge.io/pull/2792/commits/36eb228bf375cd3d840fb10a56787ba4023afd05 by moving these functions under `MqttSchema`, which defines the topic structures.

Partially fixed, because these functions are still not checking the topic root. This is done on purpose, as the point of these functions is to extract data from a topic, *even* if not actually part of subscription.